pkg/metrics/sketch_series.go

```diff
@@ -0,0 +1,94 @@
+package metrics
+
+import (
+	"errors"
+
+	"github.com/DataDog/agent-payload/gogen"
+	"github.com/DataDog/datadog-agent/pkg/aggregator/ckey"
+	"github.com/DataDog/datadog-agent/pkg/quantile"
+	"github.com/DataDog/datadog-agent/pkg/serializer/marshaler"
+)
+
+// A SketchSeries is a timeseries of quantile sketches.
+type SketchSeries struct {
+	Name       string
+	Tags       []string
+	Host       string
+	Interval   int64
+	Points     []SketchPoint
+	ContextKey ckey.ContextKey
+}
+
+// A SketchPoint represents a quantile sketch at a specific time
+type SketchPoint struct {
+	Sketch *quantile.Sketch
+	Ts     int64
+}
+
+// A SketchSeriesList implements marshaler.Marshaler
+type SketchSeriesList []SketchSeries
+
+// MarshalJSON is not supported for sketches.
+func (SketchSeriesList) MarshalJSON() ([]byte, error) {
+	return nil, errors.New("sketches don't support json encoding")
+}
+
+// Marshal encodes this series list.
+func (sl SketchSeriesList) Marshal() ([]byte, error) {
+	pb := &gogen.SketchPayload{
+		Sketches: make([]gogen.SketchPayload_Sketch, 0, len(sl)),
+	}
+
+	for _, ss := range sl {
+		dsl := make([]gogen.SketchPayload_Sketch_Dogsketch, 0, len(ss.Points))
+
+		for _, p := range ss.Points {
+			b := p.Sketch.Basic
+			k, n := p.Sketch.Cols()
+			dsl = append(dsl, gogen.SketchPayload_Sketch_Dogsketch{
+				Ts:  p.Ts,
+				Cnt: b.Cnt,
+				Min: b.Min,
+				Max: b.Max,
+				Avg: b.Avg,
+				Sum: b.Sum,
+				K:   k,
+				N:   n,
+			})
+		}
+
+		pb.Sketches = append(pb.Sketches, gogen.SketchPayload_Sketch{
+			Metric:      ss.Name,
+			Host:        ss.Host,
+			Tags:        ss.Tags,
+			Dogsketches: dsl,
+		})
+	}
+
+	return pb.Marshal()
+}
+
+// SplitPayload breaks the payload into times number of pieces
+func (sl SketchSeriesList) SplitPayload(times int) ([]marshaler.Marshaler, error) {
```

Comment:
Copied this over from the last version, but that seems like a good idea, I'll take a look.

Reply:
The `Marshaler` interface makes this hard to do well for sketches because:

1. sketches sizes are not uniform which makes splitting x times not make much sense.
2. they are much much larger than normal metric points which makes repeatedly re-compressing really slow (which is made worse by all of the allocation in `split.Payloads`)
3. We don't support json

It would be much easier for us to be given a `maxPayloadSize` and `goalPayloadSize` and just pass on a `[][]byte` and just do this manually in https://github.com/DataDog/datadog-agent/blob/45fe8409a7569dac247591ae3ea34bc27f7a7d01/pkg/serializer/serializer.go#L155-L165

Reply:
The main question behind is: are you also impacted by the max payload size from the intake ?
We split payload to be under that max size and avoid dropping metrics for that. And we need to have every points for the same metric in on payload for full_cardinality_tagging (ie container related).

If your not impacted by this the split get much much simpler.

Reply:
Gonna hold of on this for a later PR, its a bit hairy to implement (and this is the same behavior as before).

Sketch timestamps are truncated to `10s` so we don't have quite the same need as metrics here

Reply:
We do not need to handle payload stripping for `full_cardinality_tagging`, sketches go through a different pipeline.

We do have a max size (that I soon want to make different than the normal points one), and may want to add some special compression of sketch internals in the future so it makes a lot of sense for this to be separate.